Sources/QuaternionModule/Quaternion.swift

```diff
@@ -0,0 +1,365 @@
+  ///
+  /// Equivalent to `Quaternion(0, imaginary)`.
+  @inlinable
+  public init(imaginary: (b: RealType, c: RealType, d: RealType)) {
+    self.init(imaginary: SIMD3(imaginary.b, imaginary.c, imaginary.d))
+  }
+
+  /// The quaternion with specified real part and imaginary parts.
+  @inlinable
+  public init(_ real: RealType, _ imaginary: SIMD3<RealType>) {
+    self.init(from: SIMD4(real, imaginary.x, imaginary.y, imaginary.z))
+  }
+
+  /// The quaternion with specified real part and imaginary parts.
+  @inlinable
+  public init(_ real: RealType, _ imaginary: (b: RealType, c: RealType, d: RealType)) {
+    self.init(real, SIMD3(imaginary.b, imaginary.c, imaginary.d))
+  }
+
+  /// The quaternion with specified real part and zero imaginary part.
+  ///
+  /// Equivalent to `Quaternion(RealType(real))`.
+  @inlinable
+  public init<Other: BinaryInteger>(_ real: Other) {
+    self.init(RealType(real))
+  }
+
+  /// The quaternion with specified real part and zero imaginary part,
+  /// if it can be constructed without rounding.
+  @inlinable
+  public init?<Other: BinaryInteger>(exactly real: Other) {
+    guard let real = RealType(exactly: real) else { return nil }
+    self.init(real)
+  }
+
+  public typealias IntegerLiteralType = Int
+
+  @inlinable
+  public init(integerLiteral value: Int) {
+    self.init(RealType(value))
+  }
+}
+
+extension Quaternion where RealType: BinaryFloatingPoint {
+  /// `other` rounded to the nearest representable value of this type.
+  @inlinable
+  public init<Other: BinaryFloatingPoint>(_ other: Quaternion<Other>) {
+    self.init(from: SIMD4(
+        RealType(other.components.x),
+        RealType(other.components.y),
+        RealType(other.components.z),
+        RealType(other.components.w)
+    ))
+  }
+
+  /// `other`, if it can be represented exactly in this type; otherwise `nil`.
+  @inlinable
+  public init?<Other: BinaryFloatingPoint>(exactly other: Quaternion<Other>) {
+    guard
+        let a = RealType(exactly: other.components.x),
+        let b = RealType(exactly: other.components.y),
+        let c = RealType(exactly: other.components.z),
+        let d = RealType(exactly: other.components.w)
+    else { return nil }
+    self.init(from: SIMD4(a, b, c, d))
+  }
+}
+
+// MARK: - Conformance to Hashable and Equatable
+extension Quaternion: Hashable {
+  @_transparent
+  public static func == (lhs: Quaternion, rhs: Quaternion) -> Bool {
+    // Identify all numbers with either component non-finite as a single "point at infinity".
+    guard lhs.isFinite || rhs.isFinite else { return true }
+    // For finite numbers, equality is defined componentwise. Cases where
+    // only one of lhs or rhs is infinite fall through to here as well, but this
+    // expression correctly returns false for them so we don't need to handle
+    // them explicitly.
+    return lhs.components == rhs.components
+  }
+
+  @_transparent
+  public func hash(into hasher: inout Hasher) {
```

Comment:
I did some tinkering and, while I am still very uncertain, I was thinking if this is maybe something that we could move to a specific `UnitQuaternion` (or `Rotation3`, or ...) – so that we'll avoid collisions on this quaternion? This would allow us to also alter the `Equatable` implementation for `UnitQuaternion` so that it will return `true` for `x == -x`.

Reply:
That said, I'll raise a PR incorporating your feedback to make x and -x hash to the same value as I am very uncertain about my suggestion.